[wip] sum

mtsmfm opened the pull request:
If we require 'active_support/core_ext/enumerable' on Ruby 2.4,
we'll see following warn because `Enumerable#sum` and `Array#sum`
are added in Ruby 2.4.

```
rails/rails/activesupport/lib/active_support/core_ext/enumerable.rb:20: warning: method redefined; discarding old sum
```

Ruby 2.4's `sum` is implemented by C lang so it's very fast
but Ruby's `sum` (`orig_sum`) assumes an `identity` is `0` when we omit `identity`
so we can delegate to `orig_sum` with explicit `identity` only.

In a strict sense, we can detect `identity` by check instance's class
but we don't have to care about that because calling `Enumerable#sum` is very rare.
In many cases we call `Array#sum` with Numeric instances.

So let's call `orig_sum` with explicit `identity` only and
suppress the warn by aliasing.

Here's the benchmark:

```
require 'benchmark/ips'

module Enumerable
  def old_sum(identity = nil, &block)
    if block_given?
      map(&block).sum(identity)
    else
      sum = identity ? inject(identity, :+) : inject(:+)
      sum || identity || 0
    end
  end

  alias :orig_sum :sum

  def new_sum(identity = nil, &block)
    if identity
      orig_sum(identity, &block)
    elsif block_given?
      map(&block).sum(identity)
    else
      inject(:+) || 0
    end
  end
end

Benchmark.ips do |x|
  x.report("old_sum without identify") { (1..10000).old_sum }
  x.report("old_sum with identify")    { (1..10000).old_sum(0) }
  x.report("new_sum without identify") { (1..10000).new_sum }
  x.report("new_sum with identify")    { (1..10000).new_sum(0) }
  x.compare!
end
```

```
Warming up --------------------------------------
old_sum without identify
                       244.000  i/100ms
old_sum with identify
                       235.000  i/100ms
new_sum without identify
                       237.000  i/100ms
new_sum with identify
                       258.854k i/100ms
Calculating -------------------------------------
old_sum without identify
                          2.349k (± 2.7%) i/s -     11.956k in   5.093771s
old_sum with identify
                          2.341k (± 2.8%) i/s -     11.750k in   5.024358s
new_sum without identify
                          2.343k (± 2.0%) i/s -     11.850k in   5.059180s
new_sum with identify
                          4.903M (± 3.8%) i/s -     24.591M in   5.023742s

Comparison:
new_sum with identify:  4903038.1 i/s
old_sum without identify:     2349.0 i/s - 2087.27x  slower
new_sum without identify:     2343.3 i/s - 2092.34x  slower
old_sum with identify:     2340.6 i/s - 2094.79x  slower
```

amatsuda commented on activesupport/lib/active_support/core_ext/enumerable.rb:
```diff
@@ -1,31 +1,61 @@
+      if identity
+        orig_sum(identity, &block)
+      else
+        if block_given?
```
`elsif`

mtsmfm replied:
Hmm, I think we should leave original `sum` as it is.

amatsuda replied:
I def agree, but where in the original can I find the

```
else
if
```

code? https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/enumerable.rb

mtsmfm replied:
L36-L42 is same as [original old sum](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/enumerable.rb#L21-L26)

amatsuda replied:
I see what you mean, but changing `else + if` to `elsif` is not really a change, and that's way better than introducing a new `else + if` IMO.

amatsuda commented on activesupport/lib/active_support/core_ext/enumerable.rb:
```diff
@@ -1,31 +1,61 @@
+        if block_given?
+          map(&block).sum(identity)
+        else
+          sum = identity ? inject(identity, :+) : inject(:+)
```
`identity` would never be `truthy here, since it's already been tested at line 33.

mtsmfm replied:
I agree with "`elsif` is not really a change" but how about this change?
Don't we have to leave original logic?

amatsuda replied:
I usually clean up such code in a separate commit (but I know some OSS maintainers tend to require PR submitters to squash atomic commits to one huge meaningless commit).

mtsmfm replied:
Ah, I intended that "we shouldn't touch and copy simply this logic I think".
However it seems that you don't think so.
I'll remove useless logics.

mtsmfm committed:
Suppress Enumerable#sum redefined warning

If we require 'active_support/core_ext/enumerable' on Ruby 2.4,
we'll see following warn because `Enumerable#sum` and `Array#sum`
are added in Ruby 2.4.

```
rails/rails/activesupport/lib/active_support/core_ext/enumerable.rb:20: warning: method redefined; discarding old sum
```

Ruby 2.4's `sum` is implemented by C lang so it's very fast
but Ruby's `sum` (`orig_sum`) assumes an `identity` is `0` when we omit `identity`
so we can delegate to `orig_sum` with explicit `identity` only.

In a strict sense, we can detect `identity` by check instance's class
but we don't have to care about that because calling `Enumerable#sum` is very rare.
In many cases we call `Array#sum` with Numeric instances.

So let's call `orig_sum` with explicit `identity` only and
suppress the warn by aliasing.

Here's the benchmark:

```
require 'benchmark/ips'

module Enumerable
  def old_sum(identity = nil, &block)
    if block_given?
      map(&block).sum(identity)
    else
      sum = identity ? inject(identity, :+) : inject(:+)
      sum || identity || 0
    end
  end

  alias :orig_sum :sum

  def new_sum(identity = nil, &block)
    if identity
      orig_sum(identity, &block)
    elsif block_given?
      map(&block).sum(identity)
    else
      inject(:+) || 0
    end
  end
end

Benchmark.ips do |x|
  x.report("old_sum without identify") { (1..10000).old_sum }
  x.report("old_sum with identify")    { (1..10000).old_sum(0) }
  x.report("new_sum without identify") { (1..10000).new_sum }
  x.report("new_sum with identify")    { (1..10000).new_sum(0) }
  x.compare!
end
```

```
Warming up --------------------------------------
old_sum without identify
                       244.000  i/100ms
old_sum with identify
                       235.000  i/100ms
new_sum without identify
                       237.000  i/100ms
new_sum with identify
                       258.854k i/100ms
Calculating -------------------------------------
old_sum without identify
                          2.349k (± 2.7%) i/s -     11.956k in   5.093771s
old_sum with identify
                          2.341k (± 2.8%) i/s -     11.750k in   5.024358s
new_sum without identify
                          2.343k (± 2.0%) i/s -     11.850k in   5.059180s
new_sum with identify
                          4.903M (± 3.8%) i/s -     24.591M in   5.023742s

Comparison:
new_sum with identify:  4903038.1 i/s
old_sum without identify:     2349.0 i/s - 2087.27x  slower
new_sum without identify:     2343.3 i/s - 2092.34x  slower
old_sum with identify:     2340.6 i/s - 2094.79x  slower
```
